aspnetcore/mvc/advanced/custom-model-binding.md

```diff
@@ -0,0 +1,98 @@
+
+You can POST a base64-encoded string to to this api method using a tool like Postman:
+
+![postman](custom-model-binding/images/postman.png "postman")
+
+The model binder will work equally well with a viewmodel instead of individual action arguments:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/ImageController.cs?name=post2&highlight=2)]
+
+As long as the binder can bind request data to appropriately named properties or arguments, the model binding will succeed.
+
+## Implementing a custom model binder
+
+You can implement your own custom model binder that converts incoming request data into keys and uses Entity Framework Core to fetch the associated entity and pass it as an argument to the action method. The simplest approach is to use the `ModelBinderAttribute`, which you apply to the type you'll be binding. With this attribute, you specify the name of the input that will be bound and the type of the `IModelBinderProvider` that it should use. The `ModelBinderAttribute` is applied to the type that should use the model binder:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Data/Author.cs?highlight=10)]
+
+The `AuthorEntityBinder` uses dependency injection to request an instance of `AppDbContext`. After confirming the expected input (in this case, "id", as specified in the attribute above) is present, it casts it to an integer and uses it to retrieve the associated entity. Finally, if successful, the result is set on the binding context. If model binding fails, the argument passed to the action method will be `null`.
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Binders/AuthorEntityBinder.cs?highlight=10)]
+
+> [!NOTE]
+> The example above specifies a default model name. This can be omitted if you always specify the `Name` on the `ModelBinderAttribute`.
+
+With this in place, action methods can use the `Author` entity directly:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/BoundAuthorsController.cs?highlight=2)]
+
+Using this approach, both the controller and action method are simplified compared to performing the same work in the controller:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/AuthorsController.cs?highlight=2)]
+
+Instead of applying an attribute, you can implement [IModelBinderProvider](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.modelbinding.imodelbinderprovider). This is how the built-in framework binders are implemented. Note that when you specify the type your binder operates on, you specify the type of argument it will produce, not the input your binder will accept. The following binder provider would work for the `AuthorEntityBinder` implemented above, and when added to MVC's collection of providers, would eliminate the need for the `ModelBinderAttribute` on `Author`.
```

Comment:
Yep, that's fine 

> I'm not showing the use of any attributes on action parameters (or on properties of models). Do you think that's important to add?

I don't think an example is needed, but it would be good to mention.